controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -0,0 +1,292 @@
+                        log.error("error encountered while trying to add entity {} for stream {}/{}", entity, scope, stream, e);
+                    } else {
+                        log.debug("entity {} added for stream {}/{} at position {}", entity, scope, stream, r);
+                    }
+                });
+    }
+
+    /**
+     * Method to remove entities from the ordered set. Entities are referred to by their position pointer.  
+     * @param scope scope
+     * @param stream stream
+     * @param entities list of entities' positions to remove
+     * @return CompletableFuture which when completed will indicate that entities are removed from the set. 
+     */
+    public CompletableFuture<Void> removeEntities(String scope, String stream, Collection<Long> entities) {
+        Set<Integer> queues = entities.stream().collect(Collectors.groupingBy(x -> new Position(x).queueNumber)).keySet();
+        return Futures.allOf(entities.stream()
+                                     .map(entity -> storeHelper.deletePath(getEntityPath(scope, stream, entity), false))
+                                     .collect(Collectors.toList()))
+                      .thenCompose(v -> Futures.allOf(queues.stream().map(queueNum -> tryDeleteSealedQueuePath(scope, stream, queueNum))
+                                                            .collect(Collectors.toList())))
+                      .whenComplete((r, e) -> {
+                          if (e != null) {
+                              log.error("error encountered while trying to remove entity positions {} for stream {}/{}", entities, scope, stream, e);
+                          } else {
+                              log.debug("entities at positions {} removed for stream {}/{}", entities, scope, stream);
+                          }
+                      });
+    }
+
+    /**
+     * Returns a map of position to entity that was added to the set. 
+     * Note: Entities are ordered by position in the set but the map responded from this api is not ordered by default.
+     * Users can filter and order elements based on the position and entity id.  
+     * @param scope scope scope
+     * @param stream stream stream
+     * @return CompletableFuture which when completed will contain all positions to entities in the set. 
+     */
+    public CompletableFuture<Map<Long, String>> getEntitiesWithPosition(String scope, String stream) {
+        Map<Long, String> result = new ConcurrentHashMap<>();
+        return Futures.exceptionallyExpecting(storeHelper.getChildren(getStreamPath(scope, stream)), DATA_NOT_FOUND_PREDICATE, Collections.emptyList())
+                          .thenCompose(children -> {
+                              // start with smallest queue and collect records
+                              Iterator<Integer> iterator = children.stream().mapToInt(Integer::parseInt).iterator();
+
+                              return Futures.loop(iterator::hasNext, () -> {
```

Comment:
There is an overload to `Futures.loop` that takes in an iterator and does the legwork for you. I've seen you using this in a few other places too.